NuMPI/Optimization/bugnicourt_cg.py

```diff
@@ -0,0 +1,242 @@
+
+    n_iterations = 1
+
+    for i in range(1, maxiter + 1):
+        sig = signature(hessp)
+        if len(sig.parameters) == 2:
+            hessp_val = hessp(x, des_dir)
+        elif len(sig.parameters) == 1:
+            hessp_val = hessp(des_dir)
+        else:
+            raise ValueError('hessp function has to take max 1 arg (descent '
+                             'dir) or 2 args (x, descent direction)')
+        denominator_temp = comm.sum(des_dir.T * hessp_val)
+
+        if denominator_temp == 0:
+            print("it {}: denominator for alpha is 0".format(i))
+
+        alpha = -comm.sum(residual.T * des_dir) / denominator_temp
+
+        if alpha < 0:
+            print("it {} : hessian is negative along the descent direction. "
+                  "You will probably need linesearch "
+                  "or trust region".format(i))
+
+        x += alpha * des_dir
+
+        '''finding new contact points and making the new_gap admissible
+        according to these contact points.'''
+
+        mask_neg = x <= bounds
+        x[mask_neg] = bounds[mask_neg]
+
+        if mean_val is not None:
+            # x = (mean_val / comm.sum(x) * nb_DOF) * x
+            x = bounds + (mean_val - mean_bounds) \
+                / (comm.sum(x) / nb_DOF - mean_bounds) * (x - bounds)
+        residual_old = residual
+
+        '''Residual = A^(-1).(U) - d A −1 .(U ) −  ∂ψadh/∂g'''
+        residual = fun(x)[1]
+
+        if mean_val is not None:
+            #
+            mask_nonzero = x > bounds
+            N_mask_nonzero = comm.sum(np.count_nonzero(mask_nonzero))
+            residual = residual \
+                - comm.sum(residual[mask_nonzero]) / N_mask_nonzero
+
+        '''Apply the admissible Lagrange multipliers.'''
+        mask_res = residual >= 0
+        mask_bounded = np.logical_and(mask_neg, mask_res)
+        residual[mask_bounded] = 0.0
+
+        if mean_val is not None:
+            #
+            mask_nonzero = residual != 0
+            N_nonzero = comm.sum(np.count_nonzero(mask_nonzero))
+            residual[mask_nonzero] = residual[mask_nonzero] - comm.sum(
+                residual[mask_nonzero]) / N_nonzero
+            # assert np.mean(residual) < 1e-14 * np.max(abs(residual))
+
+        '''Computing beta for updating descent direction
+            beta = num / denom
+            num = new_residual_transpose . (new_residual - old_residual)
+            denom = alpha * descent_dir_transpose . (A_inverse - d2_ψadh).
+            descent_dir '''
+
+        # beta = np.sum(residual.T * hessp_val) / denominator_temp
+        beta = comm.sum(residual * (residual - residual_old)) / (
+                alpha * denominator_temp)
+
+        des_dir_old = des_dir
+        des_dir = -residual + beta * des_dir_old
+
+        # TODO: this is useless
+        des_dir[np.logical_not(mask_bounded)] = -residual[
+            np.logical_not(mask_bounded)] + beta * des_dir_old[
+            np.logical_not(mask_bounded)]
```

Comment:
@sindhu2993 . I wrote this is useless because des_dir[mak_bounded] is set to zero just below.

- Can you confirm this is not necessary ?
- Do you think what you do is improving performance ? @pastewka  do you have a clue ?

Reply:
okay so the test work without it as well. I will comment this out.